### PR TITLE
gitless 0.8.3 initial submission

### PR DESCRIPTION
**Pull request by knowsuchagency**
I have read the submission guidelines and tried to ensure the pull request follows code style guidelines and passes tests.

I have checked to see that no one else has submitted a pull request for gitless.

The formula seems to install properly and passes tests.

The formula has passed audits using the flags `new-formula`, `online`, and `strict`.

Please let me know if there is anything else I can do to help get this merged.
## Thanks!

gitless is a CLI that wraps git and provides an alternative
workflow without the concept of a staging area.

This formula downloads an already compiled binary and simply moves it
to the bin directory.

### CODE REVIEW

File: Formula/gitless.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class Gitless < Formula
+  desc "Gitless: a version control system"
+  homepage "http://gitless.com/"
+  url "https://github.com/sdg-mit/gitless/releases/download/v0.8.3/gl-v0.8.3-darwin-x86_64.tar.gz"
+  version "0.8.3"
+  sha256 "4b5cdc00da5fe93a12904c9992a332484201b979687ce9cce5160576c7cb2048"
+
+  def install
+    # The tarball is already compiled.
+    # No need to do anything but move it.
```
Comment (MikeMcQuaid) [inline Formula/gitless.rb]: Can remove these comments.

File: Formula/gitless.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class Gitless < Formula
+  desc "Gitless: a version control system"
+  homepage "http://gitless.com/"
+  url "https://github.com/sdg-mit/gitless/releases/download/v0.8.3/gl-v0.8.3-darwin-x86_64.tar.gz"
```
Comment (MikeMcQuaid) [inline Formula/gitless.rb]: We need to use a source tarball and build from source.

File: Formula/gitless.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  def install
+    # The tarball is already compiled.
+    # No need to do anything but move it.
+    bin.install "gl"
+  end
+
+  test do
+    # The output of `gl init` has non-ascii characters and will
+    # cause the test to fail, so we revert to `gl version`.
+    system "#{bin}/gl" " --version"
```
Comment (MikeMcQuaid) [inline Formula/gitless.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

File: Formula/gitless.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+  def install
+    # The tarball is already compiled.
+    # No need to do anything but move it.
+    bin.install "gl"
+  end
+
+  test do
+    # The output of `gl init` has non-ascii characters and will
+    # cause the test to fail, so we revert to `gl version`.
```
Comment (MikeMcQuaid) [inline Formula/gitless.rb]: Remove this comment

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by knowsuchagency** [comment]
Thanks for the feedback, Mike. I went ahead and read the [Python Formula Cookbook](https://github.com/Homebrew/brew/blob/master/docs/Python-for-Formula-Authors.md) to try and learn how to properly write the formula.

In particular, I found the [poet](https://pypi.python.org/pypi/homebrew-pypi-poet) library especially useful.

After modifying the output of the poet, this is formula I came up with for gitless:
```ruby
class Gitless < Formula

  desc "Gitless: a version control system"
  homepage "http://gitless.com/"
  url "https://github.com/scottsideleau/gitless/archive/master.tar.gz"
  version "0.8.3"
  sha256 "df212d6a3d377f90c46d03fd4216ecbbeb40055b62c0fe73b9c3aa165ea8f0af"

  depends_on :python if MacOS.version <= :snow_leopard
  depends_on "libgit2"

  resource "args" do
    url "https://files.pythonhosted.org/packages/e5/1c/b701b3f4bd8d3667df8342f311b3efaeab86078a840fb826bd204118cc6b/args-0.1.0.tar.gz"
    sha256 "a785b8d837625e9b61c39108532d95b85274acd679693b71ebb5156848fcf814"
  end

  resource "cffi" do
    url "https://files.pythonhosted.org/packages/0a/f3/686af8873b70028fccf67b15c78fd4e4667a3da995007afc71e786d61b0a/cffi-1.8.3.tar.gz"
    sha256 "c321bd46faa7847261b89c0469569530cad5a41976bb6dba8202c0159f476568"
  end

  resource "clint" do
    url "https://files.pythonhosted.org/packages/3d/b4/41ecb1516f1ba728f39ee7062b9dac1352d39823f513bb6f9e8aeb86e26d/clint-0.5.1.tar.gz"
    sha256 "05224c32b1075563d0b16d0015faaf9da43aa214e4a2140e51f08789e7a4c5aa"
  end

  resource "pycparser" do
    url "https://files.pythonhosted.org/packages/6d/31/666614af3db0acf377876d48688c5d334b6e493b96d21aa7d332169bee50/pycparser-2.14.tar.gz"
    sha256 "7959b4a74abdc27b312fed1c21e6caf9309ce0b29ea86b591fd2e99ecdf27f73"
  end

  resource "pygit2" do
    url "https://files.pythonhosted.org/packages/aa/56/84dcce942a48d4b7b970cfb7a779b8db1d904e5ec5f71e7a67a63a23a4e2/pygit2-0.24.1.tar.gz"
    sha256 "4d1d0196b38d6012faf0a7c45e235c208315672b6035da504566c605ba494064"
  end

  resource "sh" do
    url "https://files.pythonhosted.org/packages/39/ca/1db6ebefdde0a7b5fb639ebc0527d8aab1cdc6119a8e4ac7c1c0cc222ec5/sh-1.11.tar.gz"
    sha256 "590fb9b84abf8b1f560df92d73d87965f1e85c6b8330f8a5f6b336b36f0559a4"
  end

  resource "six" do
    url "https://files.pythonhosted.org/packages/b3/b2/238e2590826bfdd113244a40d9d3eb26918bd798fc187e2360a8367068db/six-1.10.0.tar.gz"
    sha256 "105f8d68616f8248e24bf0e9372ef04d3cc10104f1980f54d57b2ce73a5ad56a"
  end

  include Language::Python::Virtualenv

  def install
    virtualenv_install_with_resources
  end

  test do
    system "#{bin}/gl" " --help"
  end
end
```

Sadly, I'm now getting installation errors that are beyond my current ability to debug.

I read somewhere that this may be due to issues with SIP on Mac OS 10.11+ that prevents homebrew from properly installing Python packages. 

As far as using a more substantial test besides the `--version` flag, it seems `--help` will work.

There is currently a bug with the way many gitless encodes its output. I will be submitting a pull request with a fix in the very near future. Once that is accepted and the other issues with installation are solved, I would certainly write a better test.

**Comment by mistydemeo** [comment]
>   url "https://github.com/scottsideleau/gitless/archive/master.tar.gz"
>   version "0.8.3"

Instead of the master tarball, you can use the tarball for that tag: https://github.com/scottsideleau/gitless/archive/v0.8.3.tar.gz

**Comment by knowsuchagency** [comment]
@mistydemeo I had originally done that but then
$`brew audit --strict --online --new-formula`
`Stable: version 0.8.3 is redundant with version scanned from URL`

What's the correct way to do it?

**Comment by benjaminfrank** [comment]
remove `version "0.8.3"` from your formula

**Comment by scottsideleau** [comment]
@knowsuchagency : Pull #45 of the sdg-mit Gitless repository has my build script changes accepted.

https://github.com/sdg-mit/gitless/pull/45

I would suggest trying to reference that version or see if @spderosso will push a new tag, based on the flurry of recently accepted pulls.

File: Formula/gitless.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+class Gitless < Formula
+  desc "Gitless: a version control system"
```
Comment (bfontaine) [inline Formula/gitless.rb]: Please don’t include the formula name in its description.

**Comment by knowsuchagency** [comment]
@scottsideleau I'm still getting the same pip installation error even with your changes, unfortunately
```
Removing source in /tmp/pip-iMUdBz-build
Command "/usr/local/Cellar/gitless/0.8.3/libexec/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-iMUdBz-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-AqYhD_-record/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/Cellar/gitless/0.8.3/libexec/include/site/python2.7/cffi" failed with error code 1 in /tmp/pip-iMUdBz-build/
Exception information:
Traceback (most recent call last):
  File "/usr/local/Cellar/gitless/0.8.3/libexec/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/local/Cellar/gitless/0.8.3/libexec/lib/python2.7/site-packages/pip/commands/install.py", line 317, in run
    prefix=options.prefix_path,
  File "/usr/local/Cellar/gitless/0.8.3/libexec/lib/python2.7/site-packages/pip/req/req_set.py", line 742, in install
    **kwargs
  File "/usr/local/Cellar/gitless/0.8.3/libexec/lib/python2.7/site-packages/pip/req/req_install.py", line 880, in install
    spinner=spinner,
  File "/usr/local/Cellar/gitless/0.8.3/libexec/lib/python2.7/site-packages/pip/utils/__init__.py", line 718, in call_subprocess
    % (command_desc, proc.returncode, cwd))
InstallationError: Command "/usr/local/Cellar/gitless/0.8.3/libexec/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-iMUdBz-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-AqYhD_-record/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/Cellar/gitless/0.8.3/libexec/include/site/python2.7/cffi" failed with error code 1 in /tmp/pip-iMUdBz-build/
```

**Comment by kontrafiktion** [comment]
I created an experimental 0.8.4 release of gitless (at https://github.com/kontrafiktion/gitless/releases/tag/0.8.4) and the following Formula works for me:

```class Gitless < Formula

  include Language::Python::Virtualenv

  desc "a simplified version control system on top of git"
  homepage "http://gitless.com/"
  url "https://github.com/kontrafiktion/gitless/archive/0.8.4.tar.gz"
  sha256 "66fb53c17e989b186d5f25aae6e8b04d38b1c1c5fdf43c54eb3e710482aacdb3"

  depends_on :python if MacOS.version <= :snow_leopard
  depends_on "libgit2"

  resource "args" do
    url "https://files.pythonhosted.org/packages/e5/1c/b701b3f4bd8d3667df8342f311b3efaeab86078a840fb826bd204118cc6b/args-0.1.0.tar.gz"
    sha256 "a785b8d837625e9b61c39108532d95b85274acd679693b71ebb5156848fcf814"
  end

  resource "cffi" do
    url "https://files.pythonhosted.org/packages/0a/f3/686af8873b70028fccf67b15c78fd4e4667a3da995007afc71e786d61b0a/cffi-1.8.3.tar.gz"
    sha256 "c321bd46faa7847261b89c0469569530cad5a41976bb6dba8202c0159f476568"
  end

  resource "clint" do
    url "https://files.pythonhosted.org/packages/3d/b4/41ecb1516f1ba728f39ee7062b9dac1352d39823f513bb6f9e8aeb86e26d/clint-0.5.1.tar.gz"
    sha256 "05224c32b1075563d0b16d0015faaf9da43aa214e4a2140e51f08789e7a4c5aa"
  end

  resource "pycparser" do
    url "https://files.pythonhosted.org/packages/6d/31/666614af3db0acf377876d48688c5d334b6e493b96d21aa7d332169bee50/pycparser-2.14.tar.gz"
    sha256 "7959b4a74abdc27b312fed1c21e6caf9309ce0b29ea86b591fd2e99ecdf27f73"
  end

  resource "pygit2" do
    url "https://files.pythonhosted.org/packages/aa/56/84dcce942a48d4b7b970cfb7a779b8db1d904e5ec5f71e7a67a63a23a4e2/pygit2-0.24.1.tar.gz"
    sha256 "4d1d0196b38d6012faf0a7c45e235c208315672b6035da504566c605ba494064"
  end

  resource "sh" do
    url "https://files.pythonhosted.org/packages/39/ca/1db6ebefdde0a7b5fb639ebc0527d8aab1cdc6119a8e4ac7c1c0cc222ec5/sh-1.11.tar.gz"
    sha256 "590fb9b84abf8b1f560df92d73d87965f1e85c6b8330f8a5f6b336b36f0559a4"
  end

  resource "six" do
    url "https://files.pythonhosted.org/packages/b3/b2/238e2590826bfdd113244a40d9d3eb26918bd798fc187e2360a8367068db/six-1.10.0.tar.gz"
    sha256 "105f8d68616f8248e24bf0e9372ef04d3cc10104f1980f54d57b2ce73a5ad56a"
  end

  def install
    virtualenv_install_with_resources
  end

  test do
    system "#{bin}/gl", " --help"
  end
end
```

so whenever the gitless projects creates a new release, we can reference that (s.a. https://github.com/sdg-mit/gitless/issues/50).

**Comment by kontrafiktion** [comment]
0.8.4 has been released: https://github.com/sdg-mit/gitless/issues/50#issuecomment-252498225

**Comment by knowsuchagency** [comment]
@kontrafiktion I tried the formula you've written but still get this error on installation:

> Sierra 10.12

```python
==> Downloading https://files.pythonhosted.org/packages/0a/f3/686af8873b70028f
Already downloaded: /Users/me/Library/Caches/Homebrew/gitless--cffi-1.8.3.tar.gz
==> /usr/local/Cellar/gitless/0.8.4/libexec/bin/pip install -v --no-deps --no-
Last 15 lines from /Users/me/Library/Logs/Homebrew/gitless/04.pip:
  Removing source in /tmp/pip-9xqSxG-build
Command "/usr/local/Cellar/gitless/0.8.4/libexec/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-9xqSxG-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-mE4iOD-record/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/Cellar/gitless/0.8.4/libexec/include/site/python2.7/cffi" failed with error code 1 in /tmp/pip-9xqSxG-build/
Exception information:
Traceback (most recent call last):
  File "/usr/local/Cellar/gitless/0.8.4/libexec/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/local/Cellar/gitless/0.8.4/libexec/lib/python2.7/site-packages/pip/commands/install.py", line 317, in run
    prefix=options.prefix_path,
  File "/usr/local/Cellar/gitless/0.8.4/libexec/lib/python2.7/site-packages/pip/req/req_set.py", line 742, in install
    **kwargs
  File "/usr/local/Cellar/gitless/0.8.4/libexec/lib/python2.7/site-packages/pip/req/req_install.py", line 880, in install
    spinner=spinner,
  File "/usr/local/Cellar/gitless/0.8.4/libexec/lib/python2.7/site-packages/pip/utils/__init__.py", line 718, in call_subprocess
    % (command_desc, proc.returncode, cwd))
InstallationError: Command "/usr/local/Cellar/gitless/0.8.4/libexec/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-9xqSxG-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-mE4iOD-record/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/Cellar/gitless/0.8.4/libexec/include/site/python2.7/cffi" failed with error code 1 in /tmp/pip-9xqSxG-build/
```

**Comment by kontrafiktion** [comment]
@knowsuchagency I have no idea. Could you please check that your versions of `virtualenv, `python`, and `pip` are up to date.
Perhaps Homebrew hides some more error information. If you are using OS X own python perhaps python/pip cannot find the "Python.h" file. What happens if you install gitless via `pip`?

**Comment by kontrafiktion** [comment]
perhaps someone else is interested in trying this out:

```class Gitless < Formula

  include Language::Python::Virtualenv

  desc "a simplified version control system on top of git"
  homepage "http://gitless.com/"
  url "https://github.com/sdg-mit/gitless/archive/v0.8.4.tar.gz"
  sha256 "ecde4887eb20109a0345a1bfee420140522f981deff335d552908a00952853c8"

  depends_on :python if MacOS.version <= :snow_leopard
  depends_on "libgit2"

  resource "args" do
    url "https://files.pythonhosted.org/packages/e5/1c/b701b3f4bd8d3667df8342f311b3efaeab86078a840fb826bd204118cc6b/args-0.1.0.tar.gz"
    sha256 "a785b8d837625e9b61c39108532d95b85274acd679693b71ebb5156848fcf814"
  end

  resource "cffi" do
    url "https://files.pythonhosted.org/packages/0a/f3/686af8873b70028fccf67b15c78fd4e4667a3da995007afc71e786d61b0a/cffi-1.8.3.tar.gz"
    sha256 "c321bd46faa7847261b89c0469569530cad5a41976bb6dba8202c0159f476568"
  end

  resource "clint" do
    url "https://files.pythonhosted.org/packages/3d/b4/41ecb1516f1ba728f39ee7062b9dac1352d39823f513bb6f9e8aeb86e26d/clint-0.5.1.tar.gz"
    sha256 "05224c32b1075563d0b16d0015faaf9da43aa214e4a2140e51f08789e7a4c5aa"
  end

  resource "pycparser" do
    url "https://files.pythonhosted.org/packages/6d/31/666614af3db0acf377876d48688c5d334b6e493b96d21aa7d332169bee50/pycparser-2.14.tar.gz"
    sha256 "7959b4a74abdc27b312fed1c21e6caf9309ce0b29ea86b591fd2e99ecdf27f73"
  end

  resource "pygit2" do
    url "https://files.pythonhosted.org/packages/aa/56/84dcce942a48d4b7b970cfb7a779b8db1d904e5ec5f71e7a67a63a23a4e2/pygit2-0.24.1.tar.gz"
    sha256 "4d1d0196b38d6012faf0a7c45e235c208315672b6035da504566c605ba494064"
  end

  resource "sh" do
    url "https://files.pythonhosted.org/packages/39/ca/1db6ebefdde0a7b5fb639ebc0527d8aab1cdc6119a8e4ac7c1c0cc222ec5/sh-1.11.tar.gz"
    sha256 "590fb9b84abf8b1f560df92d73d87965f1e85c6b8330f8a5f6b336b36f0559a4"
  end

  resource "six" do
    url "https://files.pythonhosted.org/packages/b3/b2/238e2590826bfdd113244a40d9d3eb26918bd798fc187e2360a8367068db/six-1.10.0.tar.gz"
    sha256 "105f8d68616f8248e24bf0e9372ef04d3cc10104f1980f54d57b2ce73a5ad56a"
  end

  def install
    virtualenv_install_with_resources
  end

  test do
    system "#{bin}/gl", " --help"
  end
end
```
or should I create a new PR? 
@knowsuchagency  I wouldn't want to do that without your permission

**Comment by kontrafiktion** [comment]
@knowsuchagency could please post the content of ~/Library/Logs/Homebrew/gitless (as a Gist)?

**Comment by knowsuchagency** [comment]
@kontrafiktion [here are the logs](https://gist.github.com/knowsuchagency/893f136d28ddfe29711f86ee46b77d6a)

I don't mind if you submit another pull request, I just want the formula to work for people on either sierra or el capitan.

**Comment by kontrafiktion** [comment]
@knowsuchagency please try

    xcode-select --install

and then try installing gitless again

**Comment by scottsideleau** [comment]
Following the 0.8.4 release, I can install gitless via Homebrew's pip3.5 for Homebrew's Python 3.5 on MacOS 10.11 (El Capitan) without error.

**Comment by kontrafiktion** [comment]
@scottsideleau I do not know enough about creating Formulas, but I found this: https://github.com/Homebrew/homebrew-core/issues/1450#issuecomment-221913342

**Comment by MikeMcQuaid** [comment]
@kontrafiktion Could you try and open a pull request? [This document](https://github.com/Homebrew/brew/blob/master/docs/How-To-Open-a-Homebrew-Pull-Request-(and-get-it-merged).md#how-to-open-a-homebrew-pull-request-and-get-it-merged) should help and I'm happy to walk you through anything else.

Thanks!

**Comment by kontrafiktion** [comment]
@MikeMcQuaid done: see #6194

**Comment by knowsuchagency** [comment]
@kontrafiktion Thanks for submitting the new pull request. I just tried installing gitless on my machine from the new formula and it worked like a charm. You da best :)

**Pull request closed without merging**